Release notes `0.26.0` (#3048)
PennyLaneAI/pennylane: doc/releases/changelog-0.26.0.md

<h3>New features since last release</h3>

<h4>Classical shadows 👤</h4>

> * All-new features for implementing the classical-shadows protocol are now available. 

Reviewer: Perhaps?
> PennyLane now provides built-in support for implementing the classical-shadows measurement protocol.

  [(#2820)](https://github.com/PennyLaneAI/pennylane/pull/2820)
  [(#2821)](https://github.com/PennyLaneAI/pennylane/pull/2821)
  [(#2871)](https://github.com/PennyLaneAI/pennylane/pull/2871)
  [(#2968)](https://github.com/PennyLaneAI/pennylane/pull/2968)
  [(#2959)](https://github.com/PennyLaneAI/pennylane/pull/2959)
  [(#2968)](https://github.com/PennyLaneAI/pennylane/pull/2968)

>   The classical-shadow measurement protocol is described in detail in the
>   [classical shadows paper](https://arxiv.org/abs/2002.08953).

Reviewer: ```suggestion
  The classical-shadow measurement protocol is described in detail in the paper
  [Predicting Many Properties of a Quantum System from Very Few Measurements](https://arxiv.org/abs/2002.08953).
```

Optional, feel free to accept/edit/decline. It felt a bit informal referring to the paper as the classical shadows paper!

  As part of the support for classical shadows in this release, two new finite-shot and fully-differentiable measurements are available: 

>   - QNodes returning `qml.classical_shadow` will return two entities: 
> 
>     + `bits`: 0 or 1 if the 1 or -1 eigenvalue is sampled, respectively
>     + `recipes`: the randomized Pauli measurements that are performed for each qubit, identified as a unique integer: 0 for Pauli X, 1 for Pauli Y, and 2 for Pauli Z.

Reviewer: ```suggestion
- QNodes returning the new measurement `qml.classical_shadow()` will return two entities;
  `bits` (0 or 1 if the 1 or -1 eigenvalue is sampled, respectively) and `recipes`
  (the randomized Pauli measurements that are performed for each qubit, labelled by integer):
```

Optional, but a bit more concise!

Reviewer 2: I prefer this solely because for some reason `recipes` isn't rendered as code upon deployment 🤔

    
    ```python
    dev = qml.device("default.qubit", wires=2, shots=3)

    @qml.qnode(dev)
    def circuit():
        qml.Hadamard(wires=0)
        qml.CNOT(wires=[0, 1])
        return qml.classical_shadow(wires=[0, 1])
    ```

    ```pycon
    >>> bits, recipes = circuit()
    >>> bits
    tensor([[0, 0],
            [1, 0],
            [0, 1]], dtype=uint8, requires_grad=True)
    >>> recipes
    tensor([[2, 2],
            [0, 2],
            [0, 2]], dtype=uint8, requires_grad=True)
    ```

>   - QNodes returning `qml.shadow_expval` yield the expectation value estimation using classical shadows:

Reviewer: ```suggestion
  - QNodes returning `qml.shadow_expval()` yield the expectation value estimation using classical shadows:
```

    ```python
    dev = qml.device("default.qubit", wires=range(2), shots=10000)

    @qml.qnode(dev)
    def circuit(x, H):
        qml.Hadamard(0)
        qml.CNOT((0,1))
        qml.RX(x, wires=0)
        return qml.shadow_expval(H)

    x = np.array(0.5, requires_grad=True) 
    H = qml.Hamiltonian(
            [1., 1.], 
            [qml.PauliZ(0) @ qml.PauliZ(1), qml.PauliX(0) @ qml.PauliX(1)]
        )  
    ```

    ```pycon
    >>> circuit(x, H), 
    tensor(1.8486, requires_grad=True) 
    >>> qml.grad(circuit)(x, H))
    -0.4797000000000001
    ```

  Fully-differentiable QNode transforms for both new classical-shadows measurements are also available via
  `qml.shadows.shadow_state` and `qml.shadows.shadow_expval`, respectively.
  
  For convenient post-processing, we've also added the ability to calculate general Renyi entropies 
  by way of the `ClassicalShadow` class' `entropy` method, which requires the wires of the subsystem of interest
  and the Renyi entropy order:

  ```pycon

>   >>> shadow = ClassicalShadow(bits, recipes)

Reviewer: This should probably have the full path (is it `qml.ClassicalShadow`?)

Reviewer 3: It is indeed.

Reviewer 2: will change!

  >>> vN_entropy = shadow.entropy(wires=[0, 3], alpha=1)
  ``` 

> <h4>Qutrits: quantum circuits for tertiary degrees of freedom ☘️</h4>

Reviewer: I can't comment on it, but:

- We can remove the `type(dev)` line (and the print statement after it)
- In standard paragraphs/text, `default.qutrit` doesn't need the quotation marks :)
- The size of the example can be reduced, by:
  - Eliminating one of the qnodes (only one is needed really)
  - Removing the separate `qutrit_function` and simply including the operations in the QNode
  - removing the import statements

  ```

  We will continue to add more and more support for qutrits in future releases.

> <h4>Simplifying just got... simpler 🫰</h4>

Reviewer: love it 😆 Although the emoji does not render for me: 
![image](https://user-images.githubusercontent.com/2959003/190291394-49e69c6b-7ced-411a-9399-12107844bb8a.png)

Reviewer: Some suggestions:

- 'The `qml.simplify` module' ➡ 'The `qml.simplify()` function'
- 'intuitive additions' ➡ do you mean improvements here?
- 'simplify or do' sounds a bit odd to me

  >>> list(circuit.tape)
  [RZ(-11.566370614359172, wires=[0]) @ RY(-11.566370614359172, wires=[0]) @ RX(-11.566370614359172, wires=[0]), probs(wires=[0])]
  ```

<h4>QNSPSA optimizer 😤</h4>

* A new optimizer called `qml.QNSPSAOptimizer` is available that implements the quantum natural simultaneous
  perturbation stochastic approximation (QNSPSA) method based on 
  [Simultaneous Perturbation Stochastic Approximation of the Quantum Fisher Information](https://quantum-journal.org/papers/q-2021-10-20-567/). 
  [(#2818)](https://github.com/PennyLaneAI/pennylane/pull/2818) 

>   `qml.QNSPSAOptimizer` can be viewed as a second-order SPSA algorithm. It requires 10 circuit 
>   executions per optimization step compared to 3 for `qml.SPSAOptimizer`.
>   The additional circuit executions are used to provide a stochastic estimation of a second-order
>   metric tensor, which often helps the optimizer to achieve faster convergence. 

Reviewer: ```suggestion
  `qml.QNSPSAOptimizer` is a second-order [SPSA algorithm](https://en.wikipedia.org/wiki/Simultaneous_perturbation_stochastic_approximation), 
  which combines the convergence power of the quantum-aware Quantum Natural Gradient
  (QNG) optimization method with the reduced quantum evaluations of SPSA methods.

  While the QNSPSA optimizer requires additional circuit executions (10 executions per
  step) compared to standard SPSA optimization (3 executions per step), these additional
  evaluations are used to provide a stochastic estimation of a second-order
  metric tensor, which often helps the optimizer to achieve faster convergence.
```

Reviewer: You could also link the demo, perhaps?

> * Added `len`, `iter` and `getitem` dunder methods to a new `CompositeOp` class to help iterate over `Sum` and `Prod` operands.

Reviewer 2: This one could be more user friendly

* The `supports_reversible_diff` device capability is unused and has been removed.
  [(#2993)](https://github.com/PennyLaneAI/pennylane/pull/2993)

<h3>Breaking changes</h3>

> * Measuring an operator that might not be hermitian as an observable now raises a warning instead of an

Reviewer 2: ```suggestion
* Measuring an operator that might not be hermitian now raises a warning instead of an
```

  error. To definitively determine whether or not an operator is hermitian, use `qml.is_hermitian`.
  [(#2960)](https://github.com/PennyLaneAI/pennylane/pull/2960)

> * PennyLane no longer supports TensorFlow `<2.3.0`. Please ensure that you're using `>= 2.3.0`

Reviewer 2: ```suggestion
* PennyLane no longer supports TensorFlow `<2.3.0`. Please ensure that you're using `>=2.3.0`.
```

  [(#2683)](https://github.com/PennyLaneAI/pennylane/pull/2683/files)

> * The `ControlledOperation` class has been removed.  This was a developer-only class, so the change should not be evident to

Reviewer 2: ```suggestion
* The `ControlledOperation` class has been removed. This was a developer-only class, so the change should not be evident to
```

* The `supports_reversible_diff` device capability is unused and has been removed.
  [(#2993)](https://github.com/PennyLaneAI/pennylane/pull/2993)

<h3>Breaking changes</h3>

* Measuring an operator that might not be hermitian now raises a warning instead of an
  error. To definitively determine whether or not an operator is hermitian, use `qml.is_hermitian`.
  [(#2960)](https://github.com/PennyLaneAI/pennylane/pull/2960)

* The `ControlledOperation` class has been removed. This was a developer-only class, so the change should not be evident to
  any users. It is replaced by `Controlled`.
  [(#2990)](https://github.com/PennyLaneAI/pennylane/pull/2990)

* The default `execute` method for the `QubitDevice` base class now calls `self.statistics`
  with an additional keyword argument `circuit`, which represents the quantum tape
  being executed.
  Any device that overrides `statistics` should edit the signature of the method to include
  the new `circuit` keyword argument.
  [(#2820)](https://github.com/PennyLaneAI/pennylane/pull/2820)

> * The `expand_matrix()` has been moved from `~/operation.py` to
>   `~/math/matrix_manipulation.py`

Reviewer: ```suggestion
* The `expand_matrix()` has been moved from `pennylane.operation` to
  `pennylane.math.matrix_manipulation`
```

  [(#3008)](https://github.com/PennyLaneAI/pennylane/pull/3008)

* `qml.grouping.utils.is_commuting` has been removed, and its Pauli word logic is now part of `qml.is_commuting`.
  [(#3033)](https://github.com/PennyLaneAI/pennylane/pull/3033)

> * `qml.is_commuting` has been moved from `pennylane/transforms/commutation_dag.py` to `pennylane/ops/functions`.

Reviewer: ```suggestion
* `qml.is_commuting` has been moved from `pennylane.transforms.commutation_dag` to `pennylane.ops.functions`.
```
